# tutorial de python basico
los titulos se ponen con "#" y cuando se esta en celda de codigo los comentarios tambien con ese
###### entre mas "#" mas pequeño sale el titulo
aca un pequeño tutorial de lo que hemos visto en python, desde lo mas basico hasta un poco mas tecnico.

## 1. variables y tipos de datos básicos

primero, variables y tipos basicos. en python no hace falta declarar el tipo de una variable ya que el lenguaje lo infiere. una variable es solo un nombre que apunta a un valor en memoria.

buenas practicas de nombres:
- usar barra baja _ para variables y funciones: ejemplo "mi_variable"
- usar mayusculas completas para constantes: ejemplo "PI = 3.142"
- elegir nombres descriptivos, evita letras sueltas como "l", "O", "I"
- para funciones, es mejor usar verbos: ejemplo "calcular_promedio()"

podemos tambien poner viñetas para organizar mejor los textos:
1. existe la lista no numerada con anidamiento
- perejil   
    + cilantro
2. tambien la lista numerada
    1. fresa
    2. naranja 
3. como tambien se puede poner cursiva y ese tipo de cosas
-  *Mensaje en cursiva*
-  **Mensaje en negrita**
- ***cursiva y negrita***



In [ ]:
# ejemplos de variables
nombre = 'Samuel'
apellido = 'Angel Cardona'
edad = 20
fecha_nacimiento = '2005-09-14'

numero = 10
pi = 3.14159
texto = 'hola mundo'
verdadero = True

print(type(numero), numero)
print(type(pi), pi)
print(type(texto), texto)
print(type(verdadero), verdadero)

# funcion con anotaciones de tipo
def saludar(persona: str) -> str:
    return f"hola, {persona}"

print(saludar(nombre))

<class 'int'> 10
<class 'float'> 3.14159
<class 'str'> hola mundo
<class 'bool'> True
hola, Samuel


importantisimo, como decia arriba python detecta los tipos de variables, aunque tambien se pueden cambiar a porposito

###### int: numeros enteros, positivos o negativos, sin decimales. ejemplo: -5, 0, 42
###### float: numeros con decimales (punto flotante). ejemplo: 3.14, -0.001, 2.0
###### bool: valores logicos True o False. internamente son enteros (True=1, False=0).
###### str: cadenas de texto delimitadas por comillas simples o dobles. ejemplo: "hola", 'python'

In [10]:
# int a float
a = 5
b = float(a)
print("int -> float:", b)

# float a int (pierde decimales)
c = 3.7
d = int(c)
print("float -> int:", d)

# numero a str
e = str(123)
print("int -> str:", e, type(e))

# str a int (debe ser un numero valido)
f = int("456")
print("str -> int:", f)

# bool a int (True=1, False=0)
print("int(True) =", int(True))
print("int(False) =", int(False))

# valores a bool (cero, cadena vacia, lista vacia son False)
print("bool(0) =", bool(0))
print("bool(3) =", bool(3))
print("bool('') =", bool(''))
print("bool('apoteosico') =", bool('apoteosico'))


int -> float: 5.0
float -> int: 3
int -> str: 123 <class 'str'>
str -> int: 456
int(True) = 1
int(False) = 0
bool(0) = False
bool(3) = True
bool('') = False
bool('apoteosico') = True


In [17]:
# booleanos simples
a = True
b = False
print("a:", a, "b:", b)

# operadores logicos
print("a and b ->", a and b)
print("a or b  ->", a or b)
print("not a   ->", not a)

a: True b: False
a and b -> False
a or b  -> True
not a   -> False


## 2. f strings
los f-strings son la forma moderna y recomendada de formatear texto y se escriben con 'f' antes de las comillas y {} para variables

In [27]:
nombre = "Samuel"
edad = 20
promedio = 4.05

mensaje = f"hola {nombre}, tienes {edad} años y tu promedio semestral es {promedio:.2f}"
print(mensaje)

# tambien puedes hacer calculos dentro
print(f"el año que viene tendras {edad + 1} años") 

var1="naranja"
var2="tangelo"
print(f"hola, soy {nombre}, y me gusta la {var1} especificamente la {var2}")

# formateo de numeros
TRM = 3892.45 # a dia 19 de septiembre
print(f"TRM: ${precio:,.2f}")  # formato con separador de miles

hola Samuel, tienes 20 años y tu promedio semestral es 4.05
el año que viene tendras 21 años
hola, soy Samuel, y me gusta la naranja especificamente la tangelo
TRM: $3,892.45


## 3. operaciones matemáticas y precedencia

se pueden hacer funciones sencillas tambien en python, ademas que tambien se puede tener en cuenta la jerarquia

In [6]:
# operaciones basicas y precedencia
x = 538
y = 132

# suma, resta, mult, division real
print(x + y, x - y, x * y, x / y)

# division entera, resto, potencia
print("division entera //", x // y)
print("resto %", x % y)
print("potencia **", x ** 2)

# ejemplos de jerarquia
print("2 + 3 * 4 =", 2 + 3 * 4)       # mult va primero -> 2 + 12
print("(2 + 3) * 4 =", (2 + 3) * 4)   # parentesis cambia el orden -> 20
print("2 ** 3 * 4 =", 2 ** 3 * 4)     # potencia antes que mult -> (8*4)
print("not True or False =", not True or False) # not se evalua antes


670 406 71016 4.075757575757576
division entera // 4
resto % 10
potencia ** 289444
2 + 3 * 4 = 14
(2 + 3) * 4 = 20
2 ** 3 * 4 = 32
not True or False = False


## 4. estructuras de datos

ahora estructuras de datos: listas, tuplas, diccionarios y conjuntos. cada una sirve para algo distinto.

### listas
listas son mutables, se definen con []


In [2]:
frutas = ['manzana', 'banana', 'cereza']
print(frutas)
frutas.append('durazno')
print('despues de append:', frutas)
frutas[1] = 'mango'
print('despues de reasignar:', frutas)
print('longitud:', len(frutas))
print('slicing:', frutas[1:3])

['manzana', 'banana', 'cereza']
despues de append: ['manzana', 'banana', 'cereza', 'durazno']
despues de reasignar: ['manzana', 'mango', 'cereza', 'durazno']
longitud: 4
slicing: ['mango', 'cereza']


### tuplas 
tuplas son inmutables, se definen con ()


In [3]:
coords = (10.0, 20.0)
print(coords, type(coords))

# conversion entre tipos
lista = list(coords)
lista.append(30)
print('lista desde tupla:', lista)
print('tupla desde lista:', tuple(lista))

# desempaquetado de tuplas
x, y = coords
print(f"coordenada x: {x}, coordenada y: {y}")

(10.0, 20.0) <class 'tuple'>
lista desde tupla: [10.0, 20.0, 30]
tupla desde lista: (10.0, 20.0, 30)
coordenada x: 10.0, coordenada y: 20.0


### diccionarios
diccionarios mapean claves a valores, se definen con {}


In [6]:
persona = {'nombre': nombre, 'apellido': apellido, 'edad': edad}
print(persona)
print('nombre:', persona['nombre'])
persona['ciudad'] = 'envigado'
print('con ciudad:', persona)
print('keys:', list(persona.keys()))
print('values:', list(persona.values()))

# acceso seguro con get()
print("telefono:", persona.get('telefono', 'no disponible'))

# iterar sobre diccionario
for clave, valor in persona.items():
    print(f"{clave}: {valor}")

{'nombre': 'Samuel', 'apellido': 'Angel Cardona', 'edad': 19}
nombre: Samuel
con ciudad: {'nombre': 'Samuel', 'apellido': 'Angel Cardona', 'edad': 19, 'ciudad': 'envigado'}
keys: ['nombre', 'apellido', 'edad', 'ciudad']
values: ['Samuel', 'Angel Cardona', 19, 'envigado']
telefono: no disponible
nombre: Samuel
apellido: Angel Cardona
edad: 19
ciudad: envigado


### conjuntos
conjuntos son colecciones sin orden y sin duplicados


In [7]:
A = {1, 2, 3, 3}
B = set([3, 4, 5])
print('A:', A)
print('B:', B)
print('union:', A | B)
print('interseccion:', A & B)
print('diferencia A-B:', A - B)
# acceso seguro con get()
print('diferencia simetrica:', A ^ B)  # elementos que no estan en ambos

# operaciones de conjuntos
print('A es subconjunto de B?', A.issubset(B))
print('tiene elementos en common?', not A.isdisjoint(B))

A: {1, 2, 3}
B: {3, 4, 5}
union: {1, 2, 3, 4, 5}
interseccion: {3}
diferencia A-B: {1, 2}
diferencia simetrica: {1, 2, 4, 5}
A es subconjunto de B? False
tiene elementos en common? True


## 5. indexación y slicing

In [ ]:
lista = ['a', 'b', 'c', 'd', 'e', 'f']

# indexacion basica
print("primer elemento:", lista[0])
print("ultimo elemento:", lista[-1])
print("penultimo:", lista[-2])

# slicing avanzado
print("primeros 3:", lista[:3])        # desde inicio hasta posicion 3
print("ultimos 3:", lista[-3:])       # ultimos 3 elementos
print("del 1 al 4:", lista[1:4])      # elementos 1, 2, 3
print("cada 2:", lista[::2])          # elementos saltando de 2 en 2
print("reversa:", lista[::-1])        # lista en orden inverso

# modificar por rangos
lista[1:3] = ['X', 'Y', 'Z']
print("modificado:", lista)

## 6. métodos de listas y strings

In [3]:
lista1 = [1, 2, 3]
lista2 = [10, 20, 30]
suma_elem = [a + b for a, b in zip(lista1, lista2)]
print('suma elemento a elemento:', suma_elem)

s = 'hola mundo'
print('upper:', s.upper())
print('split:', s.split())

L = [5, 2, 9, 1]
print('sorted:', sorted(L))
L.sort()
print('lista ordenada:', l)
L[2] = 99
print("cambio index 2 ->", L)
L[1:3] = [10,11]   # cambiar por rango
print("reemplazo rango 1:3 ->", L)
L.append(5)        # agregar al final
print("append ->", L)
L.insert(1, 'x')   # insertar en posicion 1
print("insert ->", L)
L.pop()            # quitar ultimo
print("pop ->", L)
L.remove('x')      # quitar primer 'x' si existe
print("remove ->", L)

suma elemento a elemento: [11, 22, 33]
upper: HOLA MUNDO
split: ['hola', 'mundo']
sorted: [1, 2, 5, 9]
lista ordenada: [1, 2, 5, 9]
cambio index 2 -> [1, 2, 99, 9]
reemplazo rango 1:3 -> [1, 10, 11, 9]
append -> [1, 10, 11, 9, 5]
insert -> [1, 'x', 10, 11, 9, 5]
pop -> [1, 'x', 10, 11, 9]
remove -> [1, 10, 11, 9]


In [8]:
cuadrados = [x**2 for x in range(10)]
print(cuadrados)

# comprensiones con condicionales
pares = [x for x in range(20) if x % 2 == 0]
print("numeros pares:", pares)

# comprension de diccionarios
cuadrados_dict = {x: x**2 for x in range(5)}
print("diccionario cuadrados:", cuadrados_dict)

gen = (x**2 for x in range(10))
print('next(gen):', next(gen))

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
numeros pares: [0, 2, 4, 6, 8, 10, 12, 14, 16, 18]
diccionario cuadrados: {0: 0, 1: 1, 2: 4, 3: 9, 4: 16}
next(gen): 0


## 7. estructuras de control

### condicionales (if/elif/else)

In [16]:
x = 7
if x < 0:
    print('negativo')
elif x == 0:
    print('cero')
else:
    print('positivo')

nombres = ['persona ficticia 1', 'samuel (yo)', 'persona ficticia 2']
edades = [101, 19, 99]
for i, (n, e) in enumerate(zip(nombres, edades), start=1):
    print(i, n, e)


positivo
1 persona ficticia 1 101
2 samuel (yo) 19
3 persona ficticia 2 99


In [5]:
promedio=4.01
semestre=5
practica_banrep=(promedio >= 4.0) & (semestre>=5)
practica_banrep

True

### bucles (for/while) 

In [12]:
numero_secreto = 9 #el numero se debe poner en el buscador de arriba
intento = None  # inicializamos la variable

print("adivina el numero secreto (entre 1 y 10):")

# el ciclo se repite mientras la condicion sea true
while intento != numero_secreto:
    intento = int(input("ingresa tu numero: "))
    
    if intento < numero_secreto:
        print("muy bajo, intenta otra vez")
    elif intento > numero_secreto:
        print("muy alto, intenta otra vez")

print("felicitaciones! adivinaste el numero")

adivina el numero secreto (entre 1 y 10):
felicitaciones! adivinaste el numero


## 8. manejo de excepciones
las excepciones son errores que ocurren durante la ejecucion.

try = intenta esto, except = si hay error haz esto

In [11]:
# ejemplo basico de manejo de excepciones
try:
    numero = int(input("ingresa un numero: "))
    resultado = 10 / numero
    print(f"10 dividido por {numero} es {resultado}")
except ValueError:
    print("error: debes ingresar un numero valido")
except ZeroDivisionError:
    print("error: no se puede dividir por cero")
except Exception as e:
    print(f"error inesperado: {e}")
finally:
    print("este bloque siempre se ejecuta")

10 dividido por 10 es 1.0
este bloque siempre se ejecuta


## 9. funciones 
como definirlas, usar argumentos, *args, **kwargs, y lambdas.

In [ ]:
def multiplicar(a, b=2):
    return a * b

print(multiplicar(3))
print(multiplicar(3, 5))

def resumen(*args, **kwargs):
    print('args:', args)
    print('kwargs:', kwargs)

resumen(1, 2, 3, nombre=nombre, edad=edad)

f = lambda x: x**2
print('lambda 4 ->', f(4))

nums = [1, 2, 3, 4]
print('map:', list(map(lambda x: x**2, nums)))
print('filter:', list(filter(lambda x: x%2==0, nums)))

6
15
args: (1, 2, 3)
kwargs: {'nombre': 'Samuel', 'edad': 19}
lambda 4 -> 16
map: [1, 4, 9, 16]
filter: [2, 4]


## 10. módulos e importaciones
tambien vimos como importar modulos. algunos ejemplos:

- "import math"
- "from math import sqrt"

In [25]:
import random as rnd #aca le ponemos un apodo o alias a random
from statistics import mean, median # de aca le decimos que vaya a statistics y saque la media y mediana

numeros = [rnd.randint(1, 100) for _ in range(7)] # aca en vez de usar random completo, usamos rnd
print('numeros:', numeros)
print('media y mediana:', mean(numeros), median(numeros))


numeros: [88, 51, 74, 41, 22, 73, 33]
media y mediana: 54.57142857142857 51


y listo, esto seria todo para un todo para un tutorial breve de python, gracias por ver :)